File: patch_tuition.js
const fs = require('fs');

function fixTuitionFile(path) {
    if (!fs.existsSync(path)) return;
    let content = fs.readFileSync(path, 'utf8');

    content = content.replace(/let currentProgressObj = stats \? stats\.currentCount : \{ count: 0, target: 9 \};/g, 
        'let currentProgressObj = stats && stats.currentCount ? stats.currentCount : { count: 0, target: 9 };');

    const searchTarget = `const targetCount = isDualBakeryLocal ? 17 : 9;
                
                
                
                let remainingForLoop = currentProgressObj.count;

                stats.allMilestones.forEach(ms => {
                    let currentTargetCount = targetCount;`;

    const replaceTarget = `const firstTargetCount = isDualBakeryLocal ? 17 : 9;
                const subTargetCount = isDualBakeryLocal ? 16 : 8;
                let isFirstCycleForThisCourse = true;
                let remainingForLoop = currentProgressObj.count;
                stats.allMilestones.forEach(ms => {
                    let currentTargetCount = isFirstCycleForThisCourse ? firstTargetCount : subTargetCount;`;

    content = content.replace(searchTarget, replaceTarget);

    // Some replacements may fail if format doesn't match exactly. Let's do it simply using replace regex.
    content = content.replace(/const targetCount = isDualBakeryLocal \? 17 : 9;\s*let remainingForLoop = currentProgressObj\.count;\s*stats\.allMilestones\.forEach\(ms => \{\s*let currentTargetCount = targetCount;/g, replaceTarget);

    content = content.replace(/remainingForLoop \-= currentTargetCount;/g, `remainingForLoop -= currentTargetCount;
                            isFirstCycleForThisCourse = false;`);

    // In courseProgressList, if we have currentProgressObj.count as TOTAL attendance,
    // we should format it as PROGRESS within cycle, as tuition_v3_old.js did in getProgressInfo.
    
    fs.writeFileSync(path, content, 'utf8');
}

fixTuitionFile('public/sejong/tuition_v3.js');
fixTuitionFile('public/sejong/tuition_v4.js');
console.log('Tuition files patched');

File: patch_tuition.py
import os
import re

filepath = "public/sejong/tuition_v3.js"
if os.path.exists(filepath):
    with open(filepath, "r", encoding="utf-8") as f:
        content = f.read()
    
    # We want to replace:
    # const targetCount = isDualBakeryLocal ? 17 : 9;
    # with:
    # const firstTargetCount = isDualBakeryLocal ? 17 : 9;
    # const subTargetCount = isDualBakeryLocal ? 16 : 8;
    # let isFirstCycleForThisCourse = true;
    
    old_target_def = "const targetCount = isDualBakeryLocal ? 17 : 9;"
    new_target_def = """const firstTargetCount = isDualBakeryLocal ? 17 : 9;
                const subTargetCount = isDualBakeryLocal ? 16 : 8;
                let isFirstCycleForThisCourse = true;"""
    
    content = content.replace(old_target_def, new_target_def)
    
    # We want to replace:
    # stats.allMilestones.forEach(ms => {
    # with:
    # stats.allMilestones.forEach(ms => {
    #     let currentTargetCount = isFirstCycleForThisCourse ? firstTargetCount : subTargetCount;
    
    old_loop_start = "stats.allMilestones.forEach(ms => {"
    new_loop_start = """stats.allMilestones.forEach(ms => {
                    let currentTargetCount = isFirstCycleForThisCourse ? firstTargetCount : subTargetCount;"""
    
    content = content.replace(old_loop_start, new_loop_start)
    
    # We want to replace all remainingForLoop -= targetCount;
    # with: remainingForLoop -= currentTargetCount; isFirstCycleForThisCourse = false;
    
    content = content.replace("remainingForLoop -= targetCount;", "remainingForLoop -= currentTargetCount;\n                            isFirstCycleForThisCourse = false;")
    
    # We also want to replace remainingForLoop >= targetCount
    content = content.replace("remainingForLoop >= targetCount", "remainingForLoop >= currentTargetCount")
    
    with open(filepath, "w", encoding="utf-8") as f:
        f.write(content)
    print("Patched tuition_v3.js")

